crates/nu-command/src/commands/all.rs

```diff
@@ -53,72 +50,56 @@ impl WholeStreamCommand for Command {
+    };
+
+    let err = Err(ShellError::labeled_error(
+        "Expected a condition",
+        "expected a condition",
+        args.call_info.name_tag.clone(),
+    ));
 
+    //This seems a little odd. Can't we have predicates with pipelines/multiple statements?
     let condition = {
-        if block.block.block.len() != 1 {
-            return Err(ShellError::labeled_error(
-                "Expected a condition",
-                "expected a condition",
-                tag,
-            ));
+        if all_args.predicate.block.block.len() != 1 {
+            return err;
         }
-        match block.block.block[0].pipelines.get(0) {
+        match all_args.predicate.block.block[0].pipelines.get(0) {
             Some(item) => match item.list.get(0) {
                 Some(ClassifiedCommand::Expr(expr)) => expr.clone(),
                 _ => {
-                    return Err(ShellError::labeled_error(
-                        "Expected a condition",
-                        "expected a condition",
-                        tag,
-                    ));
+                    return err;
                 }
             },
             None => {
-                return Err(ShellError::labeled_error(
-                    "Expected a condition",
-                    "expected a condition",
-                    tag,
-                ));
+                return err;
             }
         }
     };
 
-    let init = Ok(InputStream::one(
-        UntaggedValue::boolean(true).into_value(&tag),
-    ));
+    let condition = condition.clone();
+    let scope = args.scope.clone();
 
-    Ok(input
-        .fold(init, move |acc, row| {
-            let condition = condition.clone();
-            let ctx = ctx.clone();
-            ctx.scope.enter_scope();
-            ctx.scope.add_vars(&block.captured.entries);
-            ctx.scope.add_var("$it", row);
-
-            let condition = evaluate_baseline_expr(&condition, &*ctx);
-            ctx.scope.exit_scope();
-
-            let curr = acc?.drain_vec();
-            let curr = curr
-                .get(0)
-                .ok_or_else(|| ShellError::unexpected("No value to check with"))?;
-            let cond = curr.as_bool()?;
-
-            match condition {
-                Ok(condition) => match condition.as_bool() {
-                    Ok(b) => Ok(InputStream::one(
-                        UntaggedValue::boolean(cond && b).into_value(&curr.tag),
-                    )),
-                    Err(e) => Err(e),
-                },
-                Err(e) => Err(e),
-            }
-        })?
-        .to_action_stream())
+    // Variables in nu are immutable. Having the same variable accross invocations
+    // of evaluate_baseline_expr does not mutate the variables and those each
+    // invocations are independent of each other!
+    scope.enter_scope();
+    scope.add_vars(&all_args.predicate.captured.entries);
+    let result: Result<bool, ShellError> = args.input.try_all(|row| {
```

Comment:
using try_all here allows to use ? inside closure, making it easier to write this

Reply:
Ahh, I missed this comment earlier.

While it makes it easier to write, it unfortunately makes it harder to read because this adds additional complexity to the iterator.

You only have a single `?`, so it's generally better to put the logic right there so that other scan see the operation and the value types that will be output.

Reply:
Short circuiting can not be implement from within the closure when using std::iter::all.

The try_x pattern can be found in std::iter::Iterator for `for_each`, `find` and `fold`. Sadly the std lib does not include the try-variants for the other iterators.

If the reader knows the std::iter functions (and the try variants), I assume he can infer what `try_all` does. I agree, relying on Inference is not good. I have added a comment.

Reply:
It's more than you're extending the iterator with more functionality. We've been trying to keep everything just using the standard iterators as they are (where possible) so that people don't have to see and understand unfamiliar methods on the iterators. There will be exceptions, but those exceptions should be used pretty globally.

Reply:
Lets agree to disagree on this point and move on :)